imsim/lsst_image.py

```diff
@@ -0,0 +1,169 @@
+                        ignore here. i.e. it won't be an error if these parameters are present.
+            logger:     If given, a logger object to log progress.
+
+        Returns:
+            xsize, ysize
+        """
+        # This is mostly a copy of the ScatteredImageType.setup function.
+        logger.debug('image %d: Building LSST_Image: image, obj = %d,%d',
+                     image_num,image_num,obj_num)
+
+        self.nobjects = self.getNObj(config, base, image_num, logger=logger)
+        logger.debug('image %d: nobj = %d',image_num,self.nobjects)
+
+        # These are allowed for LSST_Image, but we don't use them here.
+        extra_ignore = [ 'image_pos', 'world_pos', 'stamp_size', 'stamp_xsize', 'stamp_ysize',
+                         'nobjects' ]
+        opt = { 'size': int , 'xsize': int , 'ysize': int, 'dtype': None,
+                'apply_vignetting': bool, 'apply_sky_gradient': bool,
+                'vignetting_data_file': str}
+        params = GetAllParams(config, base, opt=opt, ignore=ignore+extra_ignore)[0]
+
+        size = params.get('size',0)
+        full_xsize = params.get('xsize',size)
+        full_ysize = params.get('ysize',size)
+
+        self.apply_vignetting = params.get('apply_vignetting', False)
+        vignetting_data_file =  params.get('vignetting_data_file', False)
+        if self.apply_vignetting:
+            self.vignetting = Vignetting(vignetting_data_file)
+
+        self.apply_sky_gradient = params.get('apply_sky_gradient', False)
+
+        if (full_xsize <= 0) or (full_ysize <= 0):
+            raise GalSimConfigError(
+                "Both image.xsize and image.ysize need to be defined and > 0.")
+
+        # If image_force_xsize and image_force_ysize were set in config, make sure it matches.
+        if ( ('image_force_xsize' in base and full_xsize != base['image_force_xsize']) or
+             ('image_force_ysize' in base and full_ysize != base['image_force_ysize']) ):
+            raise GalSimConfigError(
+                "Unable to reconcile required image xsize and ysize with provided "
+                "xsize=%d, ysize=%d, "%(full_xsize,full_ysize))
+
+        return full_xsize, full_ysize
+
+    def addNoise(self, image, config, base, image_num, obj_num, current_var, logger):
+        """Add the final noise to a Scattered image
+
+        Parameters:
+            image:          The image onto which to add the noise.
+            config:         The configuration dict for the image field.
+            base:           The base configuration dict.
+            image_num:      The current image number.
+            obj_num:        The first object number in the image.
+            current_var:    The current noise variance in each postage stamps.
+            logger:         If given, a logger object to log progress.
+        """
+        base['current_noise_image'] = base['current_image']
+        sky = GetSky(config, base, full=True)
```

Comment:
I'm not sure that this necessarily returns an image (which is assumed below).  If the wcs is uniform, then the sky level is constant, and I think this will just return that constant as a float, not an image full of this value, even with full=True.  (Probably bad API design, TBH -- I suspect it would have been better to let this parameter mean always return a full image.)  
Anyway, I think we need a check below if doing the vignetting or gradient to turn the sky into an image if it isn't one already.

Reply:
Update -- I looked into this more, and there would be other inefficiencies if this always returned an image.  The API is such that with full=True, then the return value can be added to an existing image to add the right sky value everywhere.  I.e. `image += sky`.  So if it's really constant, there is no point in the extra array allocation.  If full=False, it just returns the value at the center, which in other contexts is sufficient (e.g. when calculating an estimate of the overall noise variance in an image).

Anyway, the upshot is that we should add the following at this point:
```
if not isinstance(sky, galsim.Image):
    sky = galsim.Image(bounds=image.bounds, wcs=image.wcs, init_value=sky)
```